File: sdk/core/retry-engine.ts
/**
 * Retry Engine for DevUtils SDK
 * Implements exponential backoff with jitter
 * Retries on: 408, 429, 500, 502, 503, 504
 */

export interface RetryConfig {
  maxRetries?: number;
  initialDelayMs?: number;
  maxDelayMs?: number;
  backoffMultiplier?: number;
  jitterFactor?: number;
}

export const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  maxRetries: 5,
  initialDelayMs: 500,
  maxDelayMs: 3000,
  backoffMultiplier: 2,
  jitterFactor: 0.1, // 10% jitter
};

// HTTP status codes that should trigger a retry
const RETRYABLE_STATUS_CODES = new Set([
  408, // Request Timeout
  429, // Too Many Requests (Rate Limited)
  500, // Internal Server Error
  502, // Bad Gateway
  503, // Service Unavailable
  504, // Gateway Timeout
]);

/**
 * Check if an error is retryable
 */
function isRetryable(error: any): boolean {
  if (!error || typeof error !== "object") {
    return false;
  }

  const status = error.status || error.statusCode;
  return RETRYABLE_STATUS_CODES.has(status);
}

/**
 * Calculate delay with exponential backoff and jitter
 */
function calculateDelay(
  attempt: number,
  config: Required<RetryConfig>,
): number {
  const exponentialDelay = Math.min(
    config.initialDelayMs * Math.pow(config.backoffMultiplier, attempt),
    config.maxDelayMs,
  );

  // Add jitter to prevent thundering herd
  const jitter = exponentialDelay * config.jitterFactor * Math.random();
  return exponentialDelay + jitter;
}

/**
 * Sleep for a given number of milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Retry a function with exponential backoff
 * @param fn - Function to retry
 * @param config - Retry configuration
 * @returns Result of the function
 * @throws Error if all retries are exhausted
 */
export async function retry<T>(
  fn: () => Promise<T>,
  config: Partial<RetryConfig> = {},
): Promise<T> {
  const finalConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: any;

  for (let attempt = 0; attempt <= finalConfig.maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      // Don't retry if not retryable
      if (!isRetryable(error)) {
        throw error;
      }

      // Don't retry if we've exhausted retries
      if (attempt === finalConfig.maxRetries) {
        throw error;
      }

      // Calculate delay and sleep
      const delay = calculateDelay(attempt, finalConfig);
      await sleep(delay);
    }
  }

  throw lastError;
}

/**
 * Retry with custom predicate
 * Useful for retrying on specific conditions beyond HTTP status codes
 */
export async function retryWithPredicate<T>(
  fn: () => Promise<T>,
  shouldRetry: (error: any, attempt: number) => boolean,
  config: Partial<RetryConfig> = {},
): Promise<T> {
  const finalConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: any;

  for (let attempt = 0; attempt <= finalConfig.maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      // Check if we should retry
      if (!shouldRetry(error, attempt)) {
        throw error;
      }

      // Don't retry if we've exhausted retries
      if (attempt === finalConfig.maxRetries) {
        throw error;
      }

      // Calculate delay and sleep
      const delay = calculateDelay(attempt, finalConfig);
      await sleep(delay);
    }
  }

  throw lastError;
}

File: sdk-python/devutils_sdk/core/retry_engine.py
"""Retry logic for DevUtils SDK"""

import asyncio
import time
from typing import Callable, Optional, TypeVar, Awaitable, Any
from dataclasses import dataclass

from .error_handler import DevUtilsError

T = TypeVar("T")


@dataclass
class RetryConfig:
    """Configuration for retry logic"""

    max_attempts: int = 3
    initial_delay: float = 1.0
    max_delay: float = 30.0
    backoff_multiplier: float = 2.0
    jitter: bool = True


def is_retryable(error: Any) -> bool:
    """Check if error is retryable"""
    if isinstance(error, DevUtilsError):
        return error.is_retryable()

    retryable_exceptions = (
        TimeoutError,
        ConnectionError,
        OSError,
    )
    return isinstance(error, retryable_exceptions)


def calculate_delay(attempt: int, config: RetryConfig) -> float:
    """Calculate delay for retry attempt"""
    delay = config.initial_delay * (config.backoff_multiplier ** attempt)
    delay = min(delay, config.max_delay)

    if config.jitter:
        import random

        delay = delay * (0.5 + random.random())

    return delay


async def sleep(ms: float) -> None:
    """Async sleep"""
    await asyncio.sleep(ms / 1000)


async def retry(
    fn: Callable[[], Awaitable[T]], config: Optional[RetryConfig] = None
) -> T:
    """
    Retry a function with exponential backoff

    Args:
        fn: Async function to retry
        config: Retry configuration

    Returns:
        Result from function

    Raises:
        DevUtilsError: If all retries fail
    """
    config = config or RetryConfig()
    last_error = None

    for attempt in range(config.max_attempts):
        try:
            return await fn()
        except Exception as e:
            last_error = e

            if not is_retryable(e):
                raise

            if attempt < config.max_attempts - 1:
                delay = calculate_delay(attempt, config)
                await sleep(delay)

    raise last_error or DevUtilsError("RETRY_FAILED", "All retry attempts failed")


async def retry_with_predicate(
    fn: Callable[[], Awaitable[T]],
    should_retry: Callable[[Any], bool],
    config: Optional[RetryConfig] = None,
) -> T:
    """
    Retry a function with custom predicate

    Args:
        fn: Async function to retry
        should_retry: Function to determine if result should be retried
        config: Retry configuration

    Returns:
        Result from function

    Raises:
        DevUtilsError: If all retries fail
    """
    config = config or RetryConfig()
    last_result = None

    for attempt in range(config.max_attempts):
        try:
            result = await fn()

            if not should_retry(result):
                return result

            last_result = result

            if attempt < config.max_attempts - 1:
                delay = calculate_delay(attempt, config)
                await sleep(delay)

        except Exception as e:
            if not is_retryable(e):
                raise

            if attempt < config.max_attempts - 1:
                delay = calculate_delay(attempt, config)
                await sleep(delay)

    return last_result or DevUtilsError("RETRY_FAILED", "All retry attempts failed")
